samza-kafka/src/main/java/org/apache/samza/system/kafka/KafkaSystemAdmin.java

```diff
@@ -0,0 +1,680 @@
+            });
+
+            loop.done();
+            return allMetadata;
+          }
+        };
+
+    Map<String, SystemStreamMetadata> result = strategy.run(fetchMetadataOperation,
+        new AbstractFunction2<Exception, ExponentialSleepStrategy.RetryLoop, BoxedUnit>() {
+          @Override
+          public BoxedUnit apply(Exception exception, ExponentialSleepStrategy.RetryLoop loop) {
+            if (loop.sleepCount() < MAX_RETRIES_ON_EXCEPTION) {
+              LOG.warn(String.format("Fetching systemstreampartition counts for: %s threw an exception. Retrying.",
+                  streamNames), exception);
+            } else {
+              LOG.error(String.format("Fetching systemstreampartition counts for: %s threw an exception.", streamNames),
+                  exception);
+              loop.done();
+              throw new SamzaException(exception);
+            }
+            return null;
+          }
+        }).get();
+
+    LOG.info("SystemStream partition counts for system {}: {}", systemName, result);
+    return result;
+  }
+
+  @Override
+  public Map<SystemStreamPartition, String> getOffsetsAfter(Map<SystemStreamPartition, String> offsets) {
+    // This is safe to do with Kafka, even if a topic is key-deduped. If the
+    // offset doesn't exist on a compacted topic, Kafka will return the first
+    // message AFTER the offset that was specified in the fetch request.
+    return offsets.entrySet()
+        .stream()
+        .collect(Collectors.toMap(Map.Entry::getKey, (entry) -> String.valueOf(Long.valueOf(entry.getValue()) + 1)));
+  }
+
+  @Override
+  public Map<String, SystemStreamMetadata> getSystemStreamMetadata(Set<String> streamNames) {
+    return getSystemStreamMetadata(streamNames,
+        new ExponentialSleepStrategy(DEFAULT_EXPONENTIAL_SLEEP_BACK_OFF_MULTIPLIER,
+            DEFAULT_EXPONENTIAL_SLEEP_INITIAL_DELAY_MS, DEFAULT_EXPONENTIAL_SLEEP_MAX_DELAY_MS));
+  }
+
+  @Override
+  public Map<SystemStreamPartition, SystemStreamMetadata.SystemStreamPartitionMetadata> getSSPMetadata(
+      Set<SystemStreamPartition> ssps) {
+
+    LOG.info("Fetching SSP metadata for: {}", ssps);
+    List<TopicPartition> topicPartitions = ssps.stream()
+        .map(ssp -> new TopicPartition(ssp.getStream(), ssp.getPartition().getPartitionId()))
+        .collect(Collectors.toList());
+
+    OffsetsMaps topicPartitionsMetadata;
+    topicPartitionsMetadata = fetchTopicPartitionsMetadata(topicPartitions);
+
+    Map<SystemStreamPartition, SystemStreamMetadata.SystemStreamPartitionMetadata> sspToSSPMetadata = new HashMap<>();
+    for (SystemStreamPartition ssp : ssps) {
+      String oldestOffset = topicPartitionsMetadata.getOldestOffsets().get(ssp);
+      String newestOffset = topicPartitionsMetadata.getNewestOffsets().get(ssp);
+      String upcomingOffset = topicPartitionsMetadata.getUpcomingOffsets().get(ssp);
+
+      sspToSSPMetadata.put(ssp,
```

Comment:
Are we guaranteed that one of these will be non-null?  I noticed we have that check internally.